pkg/plugins/runtime/k8s/controllers/pod_converter.go

```diff
@@ -182,6 +192,61 @@ func GatewayFor(clusterName string, pod *kube_core.Pod, services []*kube_core.Se
 	}, nil
 }
 
+// DeploymentFor returns the name of the deployment that the pod exists within. The second return
+// value indicates whether or not the deployment was found when no error occurs, otherwise an
+// error is returned as the third return value.
+func (p *PodConverter) DeploymentFor(ctx context.Context, namespace string, pod *kube_core.Pod) (string, bool, error) {
+	owners := pod.GetObjectMeta().GetOwnerReferences()
+	rs := &kube_apps.ReplicaSet{}
+	for _, owner := range owners {
+		if owner.Kind == "ReplicaSet" {
+			rsKey := kube_client.ObjectKey{Namespace: namespace, Name: owner.Name}
+			if err := p.ReplicaSetGetter.Get(ctx, rsKey, rs); err != nil {
+				return "", false, err
+			}
+			break
+		}
+	}
+
+	if rs == nil {
+		return "", false, nil
+	}
+
+	rsOwners := rs.GetObjectMeta().GetOwnerReferences()
+	dpl := &kube_apps.Deployment{}
+	for _, owner := range rsOwners {
+		if owner.Kind == "Deployment" {
+			dplKey := kube_client.ObjectKey{Namespace: namespace, Name: owner.Name}
+			if err := p.DeploymentGetter.Get(ctx, dplKey, dpl); err != nil {
+				return "", false, err
+			}
+			break
+		}
+	}
+
+	if dpl != nil {
+		return dpl.Name, true, nil
+	}
+
+	return "", false, nil
+}
+
+func (p *PodConverter) GatewayByDeploymentFor(ctx context.Context, clusterName string, pod *kube_core.Pod, services []*kube_core.Service) (*mesh_proto.Dataplane_Networking_Gateway, error) {
+	namespace := pod.GetObjectMeta().GetNamespace()
+	deployment, found, err := p.DeploymentFor(ctx, namespace, pod)
```

Comment:
you might be able to do simpler like we've done for pods without services attached: https://github.com/kumahq/kuma/blob/060d45d61cd311c3424decd53dcf4da673e66f9c/pkg/plugins/runtime/k8s/controllers/inbound_converter.go#L88

It uses naming of the pod:

https://github.com/kumahq/kuma/blob/060d45d61cd311c3424decd53dcf4da673e66f9c/pkg/plugins/runtime/k8s/controllers/inbound_converter.go#L217-L225

Reply:
Isn't that unnecessarily fragile? Owner references are structured, semantic data. The structure of the name of a pod isn't really part of the k8s API.

Reply:
The other one is just as fragile because pods can be created in many other ways (statefulsets won't have replicasets and deployments for example).
Not sure what the right way is though :)

Reply:
That's not fragility, we can branch on whether it's a replicaset and make a different choice if it isn't. The name is strictly worse because it assumes a replicaset, assumes how a replicaset names its pods and if the assumptions are wrong it just does... undefined.

Reply:
Agree pulling structured data is better than parsing Pod name.

Would it make sense to use the owner reference name as long as it is a relevant kind?

FWIW, if the pod isn't owned by a replicaset, in turn by a deployment, this currently falls back on existing service-tag behavior.

Reply:
Ok makes sense! Though we should make serviceless pods leverage this code too in the future.